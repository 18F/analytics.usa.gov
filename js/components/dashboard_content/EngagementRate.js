import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the engagement rate report from the passed data URL and creates a
 * visualization for the engagement of users visiting sites for the current
 * agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function EngagementRate({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/engagement-rate-30-days.json`;
  const ref = useRef(null);
  const [engagementRateData, setEngagementRateData] = useState(null);

  useEffect(() => {
    const initEngagementRateChart = async () => {
      if (!engagementRateData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setEngagementRateData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(engagementRateData)
          .setRenderer((selection, data) => {
            const metrics = data.data[0] || {};
            const engagementRate = parseFloat(metrics.engagementRate) * 100;
            selection.text(formatters.floatToPercent(engagementRate));
          })
          .build();
      }
    };
    initEngagementRateChart().catch(console.error);
  }, [engagementRateData]);

  return (
    <>
      <div ref={ref}>
        <div className="data"></div>
      </div>
    </>
  );
}

EngagementRate.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default EngagementRate;
