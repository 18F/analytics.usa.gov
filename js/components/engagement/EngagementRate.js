import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import formatters from "../../lib/chart_helpers/formatters";

function EngagementRate({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/engagement-rate-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    // The engagement rate is rounded and formatted to a percentage.
    const initEngagementRateChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.loadAndRender().render((selection, data) => {
            const metrics = data.data[0];
            const engagementRate = parseFloat(metrics.engagementRate) * 100;
            selection.text(formatters.floatToPercent(engagementRate));
          }),
        );
      return result;
    };
    initEngagementRateChart().catch(console.error);
  }, []);

  return (
    <>
      <div ref={ref}>
        <div className="data"></div>
      </div>
    </>
  );
}

EngagementRate.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default EngagementRate;
