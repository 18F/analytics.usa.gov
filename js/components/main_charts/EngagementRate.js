import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the engagement rate report from the passed data URL and creates a
 * visualization for the engagement of users visiting sites for the current
 * agency.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 */
function EngagementRate({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/engagement-rate-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    // The engagement rate is rounded and formatted to a percentage.
    const initEngagementRateChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.loadAndRender().render((selection, data) => {
            const metrics = data.data[0];
            const engagementRate = parseFloat(metrics.engagementRate) * 100;
            selection.text(formatters.floatToPercent(engagementRate));
          }),
        );
      return result;
    };
    initEngagementRateChart().catch(console.error);
  }, []);

  return (
    <>
      <div ref={ref}>
        <div className="data"></div>
      </div>
    </>
  );
}

EngagementRate.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default EngagementRate;
