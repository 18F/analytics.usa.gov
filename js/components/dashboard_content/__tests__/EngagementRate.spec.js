import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import DataLoader from "../../../lib/data_loader";
import EngagementRate from "../EngagementRate";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadDailyReportJSON: jest.fn(),
}));

describe("EngagementRate", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <EngagementRate dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "engagement-rate",
        query: {
          metrics: [
            {
              name: "engagementRate",
            },
          ],
          dateRanges: [
            {
              startDate: "30daysAgo",
              endDate: "yesterday",
            },
          ],
          limit: "10000",
          property: "properties/397708109",
        },
        meta: {
          name: "Engagement rate (30 Days)",
          description:
            "Percent of sessions where the user was engaged over the past 30 days.",
        },
        data: [
          {
            engagementRate: "0.567412736435461",
          },
        ],
        totals: {},
        taken_at: "2024-03-20T02:14:07.426Z",
      };

      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <EngagementRate dataHrefBase="http://www.example.com/data/" />,
      );
      await waitFor(() => screen.getByText("56.7%"));
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <EngagementRate dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
