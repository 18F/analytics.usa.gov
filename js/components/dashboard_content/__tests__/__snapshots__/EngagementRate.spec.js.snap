// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngagementRate when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data"
    >
      56.7%
    </div>
  </div>
</DocumentFragment>
`;

exports[`EngagementRate when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data"
    />
  </div>
</DocumentFragment>
`;

exports[`EngagementRate when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data"
    />
  </div>
</DocumentFragment>
`;
